Frequent timeout when running test cases in parallel

**Please state your issue using the following template and, most importantly, in English.**

**Describe the bug**
using pytest-xdist plugin, running with param `-n 4` and make sure each worker running on each collection (collection name was 
 generated by random)
some level-1 test cases executed failed caused by time execeed(360s)

**Steps/Code to reproduce behavior**
```
http://internal.zilliz.com:18080/jenkins/job/milvus-ci/job/PR-3725/13/artifact/_artifacts/centos7/gpu/milvus_gpu_mysql_dev_test.log
```
**Expected behavior**
A clear and concise description of what you expected to happen.

**Method of installation**
- [ ] Docker/cpu
- [x] Docker/gpu
- [ ] Build from source

**Environment details**
- Hardware/Softwars conditions (OS, CPU, GPU, Memory)


- Milvus version (master or released version)


**Configuration file**
Settings you made in `server_config.yaml` or `milvus.yaml`

```yaml
paste-file-content-here
```

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Additional context**
Add any other context about the problem here.

`http://internal.zilliz.com:18080/jenkins/job/milvus-ci/job/PR-3725/22/artifact/_artifacts/centos7/cpu/milvus_cpu_mysql_dev_test.log`
```
________________ TestCompactBase.test_compact_multi_collections ________________
[gw3] linux -- Python 3.6.8 /usr/local/bin/python3

self = <test_compact.TestCompactBase object at 0x7fc4b7728b38>
connect = <milvus.client.stub.Milvus object at 0x7fc4b77b71d0>

    @pytest.mark.timeout(COMPACT_TIMEOUT)
    def test_compact_multi_collections(self, connect):
        '''
        target: test compact works or not with multiple collections
        method: create 50 collections, add entities into them and compact in turn
        expected: status ok
        '''
        nb = 100
        num_collections = 20
        entities = gen_entities(nb)
        collection_list = []
        for i in range(num_collections):
            collection_name = gen_unique_str("test_compact_multi_collection_%d" % i)
            collection_list.append(collection_name)
>           connect.create_collection(collection_name, default_fields)

test_compact.py:366: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/local/lib/python3.6/site-packages/milvus/client/stub.py:35: in inner
    return func(self, *args, **kwargs)
/usr/local/lib/python3.6/site-packages/milvus/client/stub.py:223: in create_collection
    return handler.create_collection(collection_name, fields, timeout)
/usr/local/lib/python3.6/site-packages/milvus/client/grpc_handler.py:59: in handler
    raise e
/usr/local/lib/python3.6/site-packages/milvus/client/grpc_handler.py:42: in handler
    return func(self, *args, **kwargs)
/usr/local/lib/python3.6/site-packages/milvus/client/grpc_handler.py:256: in create_collection
    status = rf.result()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <_MultiThreadedRendezvous of RPC that terminated with:
	status = StatusCode.DEADLINE_EXCEEDED
	details = "Deadline Exc...31:19530","file":"src/core/lib/surface/call.cc","file_line":1062,"grpc_message":"Deadline Exceeded","grpc_status":4}"
>
timeout = None

    def result(self, timeout=None):
        """Returns the result of the computation or raises its exception.
    
        See grpc.Future.result for the full API contract.
        """
        with self._state.condition:
            timed_out = _common.wait(self._state.condition.wait,
                                     self._is_complete,
                                     timeout=timeout)
            if timed_out:
                raise grpc.FutureTimeoutError()
            else:
                if self._state.code is grpc.StatusCode.OK:
                    return self._state.response
                elif self._state.cancelled:
                    raise grpc.FutureCancelledError()
                else:
>                   raise self
E                   grpc._channel._MultiThreadedRendezvous: <_MultiThreadedRendezvous of RPC that terminated with:
E                   	status = StatusCode.DEADLINE_EXCEEDED
E                   	details = "Deadline Exceeded"
E                   	debug_error_string = "{"created":"@1602315137.143384308","description":"Error received from peer ipv4:10.106.153.131:19530","file":"src/core/lib/surface/call.cc","file_line":1062,"grpc_message":"Deadline Exceeded","grpc_status":4}"
E                   >

/usr/local/lib/python3.6/site-packages/grpc/_channel.py:625: _MultiThreadedRendezvous
----------------------------- Captured stderr call -----------------------------
RPC error: <_MultiThreadedRendezvous of RPC that terminated with:
	status = StatusCode.DEADLINE_EXCEEDED
	details = "Deadline Exceeded"
	debug_error_string = "{"created":"@1602315137.143384308","description":"Error received from peer ipv4:10.106.153.131:19530","file":"src/core/lib/surface/call.cc","file_line":1062,"grpc_message":"Deadline Exceeded","grpc_status":4}"
>
	{'API start': '2020-10-10 07:31:47.142088', 'RPC start': '2020-10-10 07:31:47.142796', 'RPC error': '2020-10-10 07:32:17.143546'}
------------------------------ Captured log call -------------------------------
[2020-10-10 07:32:17-ERROR-milvus.client.grpc_handler]: RPC error: <_MultiThreadedRendezvous of RPC that terminated with:
	status = StatusCode.DEADLINE_EXCEEDED
	details = "Deadline Exceeded"
	debug_error_string = "{"created":"@1602315137.143384308","description":"Error received from peer ipv4:10.106.153.131:19530","file":"src/core/lib/surface/call.cc","file_line":1062,"grpc_message":"Deadline Exceeded","grpc_status":4}"
>
	{'API start': '2020-10-10 07:31:47.142088', 'RPC start': '2020-10-10 07:31:47.142796', 'RPC error': '2020-10-10 07:32:17.143546'} (grpc_handler.py:58)
```

http://internal.zilliz.com:18080/jenkins/job/milvus-ci/job/PR-4035/4/artifact/_artifacts/centos7/gpu/milvus_gpu_mysql_dev_test.log
```
=================================== FAILURES ===================================
_______________ TestInsertBinary.test_insert_binary_multi_times ________________
[gw0] linux -- Python 3.6.8 /usr/local/bin/python3

self = <test_insert.TestInsertBinary object at 0x7fee34696710>
connect = <milvus.client.stub.Milvus object at 0x7fee82e846a0>
binary_collection = 'test_75Grcsxg'

    def test_insert_binary_multi_times(self, connect, binary_collection):
        '''
        target: test insert entities multi times and final flush
        method: create collection and insert binary entity multi and final flush
        expected: the collection row count equals to nb
        '''
        for i in range(default_nb):
>           ids = connect.insert(binary_collection, default_binary_entity)

entity/test_insert.py:588: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/local/lib/python3.6/site-packages/milvus/client/stub.py:33: in inner
    return func(self, *args, **kwargs)
/usr/local/lib/python3.6/site-packages/milvus/client/stub.py:351: in insert
    return handler.insert(collection_name, entities, ids, partition_tag, params, timeout, **kwargs)
/usr/local/lib/python3.6/site-packages/milvus/client/grpc_handler.py:42: in handler
    return func(self, *args, **kwargs)
/usr/local/lib/python3.6/site-packages/milvus/client/grpc_handler.py:363: in insert
    response = rf.result()
/usr/local/lib/python3.6/site-packages/grpc/_channel.py:616: in result
    timeout=timeout)
/usr/local/lib/python3.6/site-packages/grpc/_common.py:141: in wait
    _wait_once(wait_fn, MAXIMUM_WAIT_TIMEOUT, spin_cb)
/usr/local/lib/python3.6/site-packages/grpc/_common.py:106: in _wait_once
    wait_fn(timeout=timeout)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Condition(<unlocked _thread.RLock object owner=0 count=0 at 0x7fee35f23d50>, 0)>
timeout = 0.1

    def wait(self, timeout=None):
        """Wait until notified or until a timeout occurs.
    
        If the calling thread has not acquired the lock when this method is
        called, a RuntimeError is raised.
    
        This method releases the underlying lock, and then blocks until it is
        awakened by a notify() or notify_all() call for the same condition
        variable in another thread, or until the optional timeout occurs. Once
        awakened or timed out, it re-acquires the lock and returns.
    
        When the timeout argument is present and not None, it should be a
        floating point number specifying a timeout for the operation in seconds
        (or fractions thereof).
    
        When the underlying lock is an RLock, it is not released using its
        release() method, since this may not actually unlock the lock when it
        was acquired multiple times recursively. Instead, an internal interface
        of the RLock class is used, which really unlocks it even when it has
        been recursively acquired several times. Another internal interface is
        then used to restore the recursion level when the lock is reacquired.
    
        """
        if not self._is_owned():
            raise RuntimeError("cannot wait on un-acquired lock")
        waiter = _allocate_lock()
        waiter.acquire()
        self._waiters.append(waiter)
        saved_state = self._release_save()
        gotit = False
        try:    # restore state no matter what (e.g., KeyboardInterrupt)
            if timeout is None:
                waiter.acquire()
                gotit = True
            else:
                if timeout > 0:
>                   gotit = waiter.acquire(True, timeout)
E                   Failed: Timeout >360.0s

/usr/local/lib/python3.6/threading.py:299: Failed
----------------------------- Captured stderr call -----------------------------

+++++++++++++++++++++++++++++++++++ Timeout ++++++++++++++++++++++++++++++++++++

~~~~~~~~~~~~~~~~~~~~ Stack of Thread-4899 (140661083662080) ~~~~~~~~~~~~~~~~~~~~
  File "/usr/local/lib/python3.6/threading.py", line 884, in _bootstrap
    self._bootstrap_inner()
  File "/usr/local/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/usr/local/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python3.6/site-packages/grpc/_channel.py", line 1138, in channel_spin
    event = state.channel.next_call_event()

~~~~~~~~~~~~~~~~~~~~~ Stack of <unknown> (140662763030272) ~~~~~~~~~~~~~~~~~~~~~
  File "/usr/local/lib/python3.6/site-packages/execnet/gateway_base.py", line 285, in _perform_spawn
    reply.run()
  File "/usr/local/lib/python3.6/site-packages/execnet/gateway_base.py", line 220, in run
    self._result = func(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/execnet/gateway_base.py", line 967, in _thread_receiver
    msg = Message.from_io(io)
  File "/usr/local/lib/python3.6/site-packages/execnet/gateway_base.py", line 432, in from_io
    header = io.read(9)  # type 1, channel 4, payload 4
  File "/usr/local/lib/python3.6/site-packages/execnet/gateway_base.py", line 400, in read
    data = self._read(numbytes - len(buf))

+++++++++++++++++++++++++++++++++++ Timeout ++++++++++++++++++++++++++++++++++++
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,7 @@ Please mark all changes in change log and use the issue from GitHub
 # Milvus 0.11.1 (TBD)
 
 ## Bug
+-   \#3793 Frequent timeout when running test cases in parallel
 -   \#4030 The configurations name is not effective after update config file and restart docker
 
 ## Feature

--- a/core/src/scheduler/TaskTable.cpp
+++ b/core/src/scheduler/TaskTable.cpp
@@ -300,8 +300,16 @@ TaskTable::TaskToExecute() {
 
 json
 TaskTable::Dump() const {
-    json ret{{"error.message", "not support yet."}};
-    return ret;
+    std::vector<json> list;
+    std::cout << "for start" << std::endl;
+
+    std::cout << "table.front: " << table_.front() << std::endl;
+    std::cout << "table.rear: " << table_.rear() << std::endl;
+    for (auto i = table_.front(); i < table_.rear(); i++) {
+        list.push_back(table_[i]->Dump());
+    }
+    std::cout << "for end" << std::endl;
+    return json{list};
 }
 
 }  // namespace scheduler

--- a/core/src/scheduler/selector/BuildIndexPass.cpp
+++ b/core/src/scheduler/selector/BuildIndexPass.cpp
@@ -10,9 +10,11 @@
 // or implied. See the License for the specific language governing permissions and limitations under the License.
 #include <fiu/fiu-local.h>
 
+#include "knowhere/index/IndexType.h"
 #include "scheduler/SchedInst.h"
 #include "scheduler/Utils.h"
 #include "scheduler/selector/BuildIndexPass.h"
+#include "scheduler/task/BuildIndexTask.h"
 #include "scheduler/tasklabel/SpecResLabel.h"
 #ifdef MILVUS_GPU_VERSION
 namespace milvus {
@@ -22,33 +24,64 @@ void
 BuildIndexPass::Init() {
     gpu_enable_ = config.gpu.enable();
     build_gpus_ = ParseGPUDevices(config.gpu.build_index_devices());
+    cpu_type_list_ = {knowhere::IndexEnum::INDEX_FAISS_BIN_IDMAP,
+                      knowhere::IndexEnum::INDEX_FAISS_BIN_IVFFLAT,
+                      knowhere::IndexEnum::INDEX_NSG,
+#ifdef MILVUS_SUPPORT_SPTAG
+                      knowhere::IndexEnum::INDEX_SPTAG_KDT_RNT,
+                      knowhere::IndexEnum::INDEX_SPTAG_BKT_RNT,
+#endif
+                      knowhere::IndexEnum::INDEX_HNSW,
+                      knowhere::IndexEnum::INDEX_ANNOY,
+                      knowhere::IndexEnum::INDEX_RHNSWFlat,
+                      knowhere::IndexEnum::INDEX_RHNSWPQ,
+                      knowhere::IndexEnum::INDEX_RHNSWSQ};
+}
+
+bool
+BuildIndexPass::isCPUIndex(const std::string& type) {
+    if (cpu_type_list_.find(type) == cpu_type_list_.end()) {
+        /* if not found in cpu_list then gpu*/
+        return false;
+    } else {
+        return true;
+    }
 }
 
 bool
 BuildIndexPass::Run(const TaskPtr& task) {
-    if (task->Type() != TaskType::BuildIndexTask) {
-        return false;
-    }
-
-    ResourcePtr res_ptr;
-    if (!gpu_enable_) {
-        LOG_SERVER_DEBUG_ << "Gpu disabled, specify cpu to build index!";
-        res_ptr = ResMgrInst::GetInstance()->GetResource("cpu");
-    } else {
-        fiu_do_on("BuildIndexPass.Run.empty_gpu_ids", build_gpus_.clear());
-        if (build_gpus_.empty()) {
-            LOG_SERVER_WARNING_ << "BuildIndexPass cannot get build index gpu!";
+    try {
+        if (task->Type() != TaskType::BuildIndexTask) {
             return false;
         }
-        LOG_SERVER_DEBUG_ << "Specify gpu" << build_gpus_[idx_] << " to build index!";
-        res_ptr = ResMgrInst::GetInstance()->GetResource(ResourceType::GPU, build_gpus_[idx_]);
-        idx_ = (idx_ + 1) % build_gpus_.size();
+
+        auto build_index_task = std::dynamic_pointer_cast<BuildIndexTask>(task);
+        auto type = build_index_task->GetIndexType();
+        LOG_SERVER_DEBUG_ << "BuildIndexPass Build index type is " << type;
+
+        ResourcePtr res_ptr;
+        if (!gpu_enable_ || isCPUIndex(type)) {
+            LOG_SERVER_DEBUG_ << "Gpu disabled or isCPUIndex, specify cpu to build index!";
+            res_ptr = ResMgrInst::GetInstance()->GetResource("cpu");
+        } else {
+            fiu_do_on("BuildIndexPass.Run.empty_gpu_ids", build_gpus_.clear());
+            if (build_gpus_.empty()) {
+                LOG_SERVER_WARNING_ << "BuildIndexPass cannot get build index gpu!";
+                return false;
+            }
+            LOG_SERVER_DEBUG_ << "Specify gpu" << build_gpus_[idx_] << " to build index!";
+            res_ptr = ResMgrInst::GetInstance()->GetResource(ResourceType::GPU, build_gpus_[idx_]);
+            idx_ = (idx_ + 1) % build_gpus_.size();
+        }
+
+        auto label = std::make_shared<SpecResLabel>(std::weak_ptr<Resource>(res_ptr));
+        task->label() = label;
+
+        return true;
+    } catch (std::exception ex) {
+        std::string str = "Milvus server encounter exception: " + std::string(ex.what());
+        return false;
     }
-
-    auto label = std::make_shared<SpecResLabel>(std::weak_ptr<Resource>(res_ptr));
-    task->label() = label;
-
-    return true;
 }
 
 void

--- a/core/src/scheduler/selector/BuildIndexPass.h
+++ b/core/src/scheduler/selector/BuildIndexPass.h
@@ -20,6 +20,7 @@
 #include <string>
 #include <thread>
 #include <unordered_map>
+#include <unordered_set>
 #include <vector>
 
 #include "config/ConfigMgr.h"

--- a/core/src/scheduler/selector/BuildIndexPass.h
+++ b/core/src/scheduler/selector/BuildIndexPass.h
@@ -44,6 +44,11 @@ class BuildIndexPass : public Pass, public ConfigObserver {
     ConfigUpdate(const std::string& name) override;
 
  private:
+    bool
+    isCPUIndex(const std::string& type);
+
+ private:
+    std::unordered_set<std::string> cpu_type_list_;
     uint64_t idx_ = 0;
     bool gpu_enable_ = false;
     std::vector<int64_t> build_gpus_;

--- a/core/src/scheduler/task/BuildIndexTask.cpp
+++ b/core/src/scheduler/task/BuildIndexTask.cpp
@@ -119,5 +119,29 @@ BuildIndexTask::OnExecute() {
     return Status::OK();
 }
 
+std::string
+BuildIndexTask::GetIndexType() {
+    auto segment_visitor = engine::SegmentVisitor::Build(snapshot_, segment_id_);
+    auto& field_visitors = segment_visitor->GetFieldVisitors();
+
+    std::vector<std::string> index_types;
+
+    for (auto& pair : field_visitors) {
+        auto& field_visitor = pair.second;
+        auto element_visitor = field_visitor->GetElementVisitor(engine::FieldElementType::FET_INDEX);
+        if (element_visitor == nullptr) {
+            continue;  // index undefined
+        }
+        auto element = element_visitor->GetElement();
+        index_types.push_back(element->GetTypeName());
+    }
+
+    if (index_types.size() != 1) {
+        throw std::runtime_error("index type size not correct." + std::to_string(index_types.size()));
+    }
+
+    return index_types[0];
+}
+
 }  // namespace scheduler
 }  // namespace milvus

--- a/core/src/scheduler/task/BuildIndexTask.cpp
+++ b/core/src/scheduler/task/BuildIndexTask.cpp
@@ -14,6 +14,7 @@
 #include <fiu/fiu-local.h>
 #include <memory>
 #include <utility>
+#include <vector>
 
 #include "db/Utils.h"
 #include "db/engine/EngineFactory.h"

--- a/core/src/scheduler/task/BuildIndexTask.h
+++ b/core/src/scheduler/task/BuildIndexTask.h
@@ -43,6 +43,9 @@ class BuildIndexTask : public Task {
     Status
     OnExecute() override;
 
+    std::string
+    GetIndexType();
+
  private:
     void
     CreateExecEngine();

